bolt-sidecar/src/builder/template.rs

```diff
@@ -35,82 +38,145 @@ impl BlockTemplate {
+                (*address, state)
+            })
+            .collect();
 
         // Update intermediate state
-        self.state_diff
-            .diffs
-            .entry(tx.recover_signer().expect("Passed validation"))
-            .and_modify(|(nonce, balance)| {
-                *nonce = nonce.saturating_sub(1);
-                *balance += max_cost;
-            });
+        for (address, diff) in address_to_diff.iter() {
+            self.state_diff
+                .diffs
+                .entry(*address)
+                .and_modify(|(nonce, balance)| {
+                    *nonce = nonce.saturating_sub(diff.transaction_count);
+                    *balance += diff.balance;
+                });
+        }
     }
 
     /// Retain removes any transactions that conflict with the given account state.
-    pub fn retain(&mut self, address: Address, mut state: AccountState) {
-        let mut indexes = Vec::new();
-
-        for (index, tx) in self.transactions.iter().enumerate() {
-            let max_cost = max_transaction_cost(tx);
-            if tx.recover_signer().expect("passed validation") == address
-                && (state.balance < max_cost || state.transaction_count > tx.nonce())
-            {
-                tracing::trace!(
-                    %address,
-                    "Removing transaction at index {} due to conflict with account state",
-                    index
-                );
-
-                indexes.push(index);
-                // Continue to the next iteration, not updating the state
-                continue;
-            }
-
-            // Update intermediary state for next transaction (if the tx was not removed)
-            state.balance -= max_cost;
-            state.transaction_count += 1;
+    pub fn retain(&mut self, address: Address, state: AccountState) {
+        let mut indexes: Vec<usize> = Vec::new();
+
+        // The pre-confirmations made by such address, and the indexes of the signed constraints
+        // in which they appear
+        let constraints_with_address: Vec<(usize, Vec<&Constraint>)> = self
+            .signed_constraints_list
+            .iter()
+            .enumerate()
+            .map(|(idx, c)| (idx, &c.message.constraints))
+            .filter(|(_idx, c)| c.iter().any(|c| c.sender == address))
+            .map(|(idx, c)| (idx, c.iter().filter(|c| c.sender == address).collect()))
+            .collect();
+
+        // For every pre-confirmation, gather the max total balance cost,
+        // and find the one with the lowest nonce
+        let (max_total_cost, min_nonce) = constraints_with_address
+            .iter()
+            .flat_map(|c| c.1.clone())
+            .fold((U256::ZERO, u64::MAX), |mut acc, c| {
+                let nonce = c.tx_decoded.nonce();
+                if nonce < acc.1 {
+                    acc.1 = nonce;
+                }
+                acc.0 += max_transaction_cost(&c.tx_decoded);
+                acc
+            });
```

Comment:
we can avoid cloning + use named variables instead of the tuple here, wdyt?

```suggestion
            .flat_map(|(_, c)| c.as_slice())
            .fold((U256::ZERO, u64::MAX), |(total_cost, min_nonce), c| {
                (
                    total_cost + max_transaction_cost(&c.tx_decoded),
                    min_nonce.min(c.tx_decoded.nonce()),
                )
            });
```

Reply:
Thanks for the cleanup!